libyul/optimiser/VarDeclPropagator.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/VarDeclPropagator.h>
+#include <libsolidity/inlineasm/AsmData.h>
+#include <libdevcore/Visitor.h>
+#include <algorithm>
+
+#include <boost/algorithm/cxx11/all_of.hpp>
+
+using namespace std;
+using namespace dev;
+using namespace dev::yul;
+
+using dev::solidity::assembly::TypedName;
+using dev::solidity::assembly::TypedNameList;
+
+void VarDeclPropagator::operator()(Block& _block)
+{
+	m_blockScopes.push(&_block);
+	ASTModifier::operator()(_block);
+	m_blockScopes.pop();
+}
+
+void VarDeclPropagator::operator()(VariableDeclaration& _varDecl)
+{
+	if (!_varDecl.value)
+	{
+		m_emptyVarDecls.push_back(&_varDecl);
+	}
+}
+
+/**
+ * Checks whether all @p identifiers are declared in the same VariableDeclaration.
+ */
+bool VarDeclPropagator::checkAllIdentifiersMatchSame(vector<Identifier> const& identifiers) const
+{
+	yulAssert(!identifiers.empty(), "");
+
+	VariableDeclaration const& varDecl = getVarDecl((*begin(identifiers)).name);
+
+	return identifiers.size() == varDecl.variables.size() && all_of(
+		begin(identifiers),
+		end(identifiers),
+		[&](Identifier const& _ident) -> bool
+		{
+			return find_if(
+				begin(varDecl.variables),
+				end(varDecl.variables),
+				[&_ident](TypedName const& typedName) { return typedName.name == _ident.name; }
+			) != end(varDecl.variables);
+		}
+	);
+}
+
+void VarDeclPropagator::operator()(Assignment& _assignment)
+{
+	yulAssert(!_assignment.variableNames.empty(), "LHS must not be empty");
+
+	// CASE-1
+	//    var x, y, z
+	//    z, x, y = RHS
+	// -->
+	//    var z, x, y = RHS
+	if (checkAllIdentifiersMatchSame(_assignment.variableNames))
+	{
+		VariableDeclaration& varDecl = getVarDecl((*begin(_assignment.variableNames)).name);
+		if (varDecl.value == nullptr)
+		{
+			*iteratorOf(_assignment) = VariableDeclaration{
+				_assignment.location,
+				recreateTypedNameList(_assignment.variableNames, varDecl.variables),
+				_assignment.value
+			};
+
+			removeVarDecl(varDecl);
```

Comment:
I think this function modifies the statement list and thus destroys the for look in the block - or am I missing something?

Reply:
the replacement line is not modifying the underlying vector, that's why I grabbed the vector's iterator and exchange the value (statement) in-place. It's different with the `removeVarDecl` though. I need to have a look whether that's safe, or how to work around it.